acvm_backend_barretenberg/src/composer.rs

```diff
@@ -167,8 +198,235 @@ impl StandardComposer {
+    // XXX: There seems to be a bug in the C++ code
+    // where it causes a `HeapAccessOutOfBound` error
+    // for certain circuit sizes.
+    //
+    // This method calls the WASM for the circuit size
+    // if an error is returned, then the circuit size is defaulted to 2^19.
+    //
+    // This method is primarily used to determine how many group
+    // elements we need from the CRS. So using 2^19 on an error
+    // should be an overestimation.
+    fn get_circuit_size(&self, constraint_system: &ConstraintSystem) -> u32 {
+        let cs_buf = constraint_system.to_bytes();
+        let cs_ptr = self.allocate(&cs_buf);
+
+        let circuit_size = self
+            .call("acir_proofs_get_total_circuit_size", &cs_ptr)
+            .into_i32();
+        let circuit_size =
+            u32::try_from(circuit_size).expect("circuit cannot have negative number of gates");
+
+        self.free(cs_ptr);
+
+        pow2ceil(circuit_size + NUM_RESERVED_GATES)
+    }
+
+    fn get_exact_circuit_size(&self, constraint_system: &ConstraintSystem) -> u32 {
+        let cs_buf = constraint_system.to_bytes();
+        let cs_ptr = self.allocate(&cs_buf);
+
+        let circuit_size = self
+            .call("acir_proofs_get_exact_circuit_size", &cs_ptr)
+            .into_i32();
+        let circuit_size =
+            u32::try_from(circuit_size).expect("circuit cannot have negative number of gates");
+
+        self.free(cs_ptr);
+
+        circuit_size
+    }
+
+    fn compute_proving_key(&self, constraint_system: &ConstraintSystem) -> Vec<u8> {
+        use super::POINTER_BYTES;
+        use wasmer::Value;
+
+        let cs_buf = constraint_system.to_bytes();
+        let cs_ptr = self.allocate(&cs_buf);
+
+        // The proving key is not actually written to this pointer.
+        // `pk_ptr_ptr` is a pointer to a pointer which holds the proving key.
+        let pk_ptr_ptr: usize = 0;
+
+        let pk_size = self
+            .call_multiple(
+                "acir_proofs_init_proving_key",
+                vec![&cs_ptr, &Value::I32(pk_ptr_ptr as i32)],
+            )
+            .value();
+        let pk_size: usize = pk_size.unwrap_i32() as usize;
+
+        // We then need to read the pointer at `pk_ptr_ptr` to get the key's location
+        // and then slice memory again at `pk_ptr` to get the proving key.
+        let pk_ptr = self.slice_memory(pk_ptr_ptr, POINTER_BYTES);
+        let pk_ptr: usize =
+            u32::from_le_bytes(pk_ptr[0..POINTER_BYTES].try_into().unwrap()) as usize;
+
+        self.slice_memory(pk_ptr, pk_size)
+    }
+
+    fn compute_verification_key(
+        &self,
+        constraint_system: &ConstraintSystem,
+        proving_key: &[u8],
+    ) -> Vec<u8> {
+        use super::POINTER_BYTES;
+        use wasmer::Value;
+
+        let circuit_size = self.get_circuit_size(constraint_system);
+        let CRS {
+            g1_data, g2_data, ..
+        } = CRS::new(circuit_size as usize);
```

Comment:
Does this actually use g1_data? Because it might be more efficient to use a G2::new instead